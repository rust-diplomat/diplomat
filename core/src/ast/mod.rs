/// As part of the macro expansion and code generation process, Diplomat
/// generates a simplified version of the Rust AST that captures special
/// types such as opaque structs, [`Box`], and [`Result`] with utilities
/// for handling such types.
pub mod attrs;
pub use attrs::Attrs;

mod methods;
pub use methods::{BorrowedParams, Method, Param, SelfParam};

mod modules;
pub use modules::{File, Module};

mod structs;
pub use structs::{OpaqueStruct, Struct};

mod enums;
pub use enums::Enum;

mod types;
pub use types::{
    CustomType, LifetimeOrigin, ModSymbol, Mutability, PathType, PrimitiveType, TypeName,
};

mod lifetimes;
pub use lifetimes::{Lifetime, LifetimeEnv, LifetimeTransitivity, NamedLifetime};

mod paths;
pub use paths::Path;

mod idents;
pub use idents::Ident;

mod docs;
pub use docs::{DocType, Docs, DocsUrlGenerator, MarkdownStyle, RustLink, RustLinkDisplay};

mod validity;
pub use validity::ValidityError;
