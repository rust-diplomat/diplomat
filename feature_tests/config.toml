lib-name = "This Will Be Ignored For Testing Purposes (Check src/lib.rs)"
unsafe_references_in_callbacks = true

[kotlin]
domain = "dev.diplomattest"

[demo-gen]
relative-js-path = "../../js/api/"