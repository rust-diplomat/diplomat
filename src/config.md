# Configuring Diplomat

Some Diplomat backends have configurable parameters to change the behavior of their output.

For instance, Kotlin requires a `domain` parameter and `lib_name` parameter. These two parameters are used to generate a package name and folder structure on generation.

## Configuration Structures
Documentation on what configuration options are available in [`config.rs` on GitHub](https://github.com/rust-diplomat/diplomat/tree/main/tool/src/config.rs).

Note that every configuration option uses `snake_case` for consistency, and that the `SharedConfig` struct is flattened. So for setting `--config shared_config.lib_name="some_value"`, this would instead be `--config lib_name="some_value"`.

## Configuration Interfaces

Configuration information can be set in four ways:

### `config.toml`

By default, Diplomat scans for a `config.toml` in the folder where `diplomat_tool` is being run. You can change the location of this folder with the `--config_file` parameter.

The structure of `config.toml` is as follows:

```toml
# Top level table specifies Shared Config settings that apply to all backends:
lib-name = "MyLibrary"

[kotlin]
# Individual tables can override Shared Config settings:
lib-name = "LibraryNameOverride"
# Along with backend specific settings:
domain = "org.myOrganization"

[demo_gen]
explicit-generation = true

[other-library-name]
some-value = 100
```

### `diplomat-tool` CLI
When running `diplomat-tool`, you may pass in the `--config` flag for each option you wish to set:

```
./diplomat-tool kotlin ./kotlin-folder --config lib_name="MyLibrary" --config kotlin.domain = "org.myOrganization"
```

`diplomat-tool` flags take priority over `config.toml`.

### `diplomat_tool::gen`

If you call `diplomat_tool::gen` manually, then you have the option of setting configuration yourself, with the `diplomat_tool::config::Config` struct.

See [Backend Structures](#backend-structures) for more on these structures.

### `#[diplomat::config(...)]`

In `lib.rs`, any top-level `mod`ule, `struct`, or `impl` block can use the `#[diplomat::config]` attribute:

```rust
#[diplomat::config(lib_name="MyLibrary")]
struct SomeConfig;

#[diplomat::config(kotlin.domain="org.myOrganization")]
mod kotlin_specific_mod;

#[diplomat::config(...)]
impl SomeConfig {

}
```

Due to a quirk of how Diplomat reads these attributes, `#[diplomat::config]` has priority over all other methods of setting configuration.