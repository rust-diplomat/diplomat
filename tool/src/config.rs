use std::{collections::HashMap, path::Path, str};

use quote::ToTokens;
use serde::{Deserialize, Serialize};
use syn::{
    parse::{Parse, ParseStream},
    Expr, Ident, Token,
};
use toml::{value::Table, Value};

use crate::{demo_gen::DemoConfig, kotlin::KotlinConfig};

#[derive(Clone, Default, Debug, Serialize, Deserialize)]
pub struct SharedConfig {
    pub lib_name: Option<String>,
}

impl SharedConfig {
    /// Quick and dirty way to tell [`set_overrides`] whether or not to copy an override from a specific language over.
    pub fn overrides_shared(name: &str) -> bool {
        // Expect the first item in the iterator to be the name of the language, so we eliminate that:
        let name: String = name.split(".").skip(1).collect();
        matches!(name.as_str(), "lib_name")
    }

    pub fn set(&mut self, key: &str, value: Value) {
        if key == "lib_name" && value.is_str() {
            self.lib_name = value.as_str().map(|v| v.to_string());
        }
    }
}

#[derive(Clone, Default, Debug, Serialize, Deserialize)]
pub struct Config {
    #[serde(flatten)]
    pub shared_config: SharedConfig,
    #[serde(rename = "kotlin")]
    pub kotlin_config: KotlinConfig,
    #[serde(rename = "demo_gen")]
    pub demo_gen_config: DemoConfig,
    /// Any language can override what's in [`SharedConfig`]. This is a structure that holds information about those specific overrides. [`Config`] will update [`SharedConfig`] based on the current language.
    #[serde(skip)]
    pub language_overrides: HashMap<String, Value>,
}

impl Config {
    pub fn set(&mut self, key: &str, value: Value) {
        if key.starts_with("kotlin.") {
            if SharedConfig::overrides_shared(key) {
                self.language_overrides.insert(key.to_string(), value);
            } else {
                self.kotlin_config.set(&key.replace("kotlin.", ""), value);
            }
        } else if key.starts_with("demo_gen.") {
            if SharedConfig::overrides_shared(key) {
                self.language_overrides.insert(key.to_string(), value);
            } else {
                self.demo_gen_config
                    .set(&key.replace("demo_gen.", ""), value);
            }
        } else if key.starts_with("nanobind.") {
            if SharedConfig::overrides_shared(key) {
                self.language_overrides.insert(key.to_string(), value);
            } // nanobind doesn't have any other config setting
        } else {
            self.shared_config.set(key, value)
        }
    }

    pub fn get_overridden(self, target_language: &str) -> Self {
        let mut out = self.clone();

        // Look for a match of language_name.some_value in a potential key.
        let m = format!("{}.", target_language);
        for (k, v) in out.language_overrides.iter() {
            if k.starts_with(&m) {
                out.shared_config.set(&k.replace(&m, ""), v.clone());
            }
        }
        out
    }

    /// Given a filepath, read TOML formatted config settings from it (and modify the current Config struct from the read)
    pub fn read_file(&mut self, path: &Path) -> Result<(), String> {
        let config_table: Table = if path.exists() {
            let file_buf = std::fs::read(path).map_err(|e| e.to_string())?;
            let s = str::from_utf8(&file_buf).map_err(|_| "Config file is not UTF8".to_string())?;
            toml::from_str(s).map_err(|_| "Config file is not valid TOML".to_string())?
        } else {
            Table::default()
        };

        for (key, value) in config_table {
            // Quick way to take config.toml from kebab to snake case.
            // This technically means that someone could also just as easily do CamelCase and have it translated,
            // but I'm not sure I want to bother writing validation code for such a scenario.
            let key = heck::AsSnakeCase(key).to_string();
            if let toml::Value::Table(t) = value {
                for (subkey, subvalue) in t {
                    let subkey = heck::AsSnakeCase(subkey).to_string();
                    self.set(&format!("{}.{}", key, subkey), subvalue);
                }
            } else {
                self.set(&key, value);
            }
        }
        Ok(())
    }

    /// Given a vector of strings with the format `config.setting = value`, modify the `Config` struct appropriately.
    pub fn read_cli_settings(&mut self, settings: Vec<String>) {
        for c in settings {
            let split = c.split_once("=");
            if let Some((key, value)) = split {
                self.set(key, toml_value_from_str(value));
            } else {
                eprintln!("Could not read {c}, expected =");
            }
        }
    }
}

pub fn toml_value_from_str(string: &str) -> toml::Value {
    let try_parse = toml::from_str::<toml::Value>(string);

    // If there's an error parsing (because clap will not parse quotes, for example), we just treat what we're passed as a string:
    if let Ok(out) = try_parse {
        out
    } else {
        toml::Value::String(string.to_string())
    }
}

#[derive(Clone, PartialEq, Eq, Hash, Debug, Serialize)]
#[non_exhaustive]
pub(crate) struct DiplomatBackendConfigAttr {
    pub key_value_pairs: Vec<DiplomatBackendConfigKeyValue>,
}

impl Parse for DiplomatBackendConfigAttr {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let list = input.parse_terminated(DiplomatBackendConfigKeyValue::parse, Token![,])?;
        let vec = list.into_iter().collect();
        Ok(Self {
            key_value_pairs: vec,
        })
    }
}

#[derive(Clone, PartialEq, Eq, Hash, Debug, Serialize)]
#[non_exhaustive]
pub(crate) struct DiplomatBackendConfigKeyValue {
    pub key: String,
    pub value: String,
}

impl Parse for DiplomatBackendConfigKeyValue {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let mut key_str: Vec<String> = Vec::new();

        loop {
            let i: Ident = input.parse()?;

            key_str.push(i.to_string());

            if input.peek(Token![.]) {
                let _period: Token![.] = input.parse()?;
            } else {
                break;
            }
        }

        let _equals: Token![=] = input.parse()?;

        let val_expr: Expr = input.parse()?;

        let value = val_expr.to_token_stream().to_string();

        Ok(Self {
            key: key_str.join("."),
            value,
        })
    }
}

pub(crate) fn find_top_level_attr(module_items: Vec<syn::Item>) -> Vec<DiplomatBackendConfigAttr> {
    let path = syn::parse_str("diplomat::config").unwrap();

    let attrs = module_items
        .iter()
        .filter_map(|i| match i {
            syn::Item::Struct(s) => Some(s.attrs.clone()),
            syn::Item::Impl(i) => Some(i.attrs.clone()),
            syn::Item::Mod(m) => Some(m.attrs.clone()),
            _ => None,
        })
        .filter_map(|attrs| {
            let attributes_vec = attrs
                .iter()
                .filter_map(|attribute| {
                    if attribute.path() == &path {
                        Some(
                            syn::parse2::<DiplomatBackendConfigAttr>(
                                attribute
                                    .parse_args()
                                    .expect("Failed to parse malformed diplomat::config"),
                            )
                            .expect("Could not parse DiplomatBackendConfig attribute."),
                        )
                    } else {
                        None
                    }
                })
                .collect::<Vec<_>>();

            if !attributes_vec.is_empty() {
                Some(attributes_vec)
            } else {
                None
            }
        });

    let mut out_config = Vec::new();

    for mut a in attrs {
        out_config.append(&mut a);
    }

    out_config
}
