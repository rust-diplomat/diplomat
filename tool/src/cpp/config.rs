use serde::Deserialize;

#[derive(Debug, Deserialize)]
pub struct CallableLibraryType {
    pub name: String,
    pub expr: String,
    pub is_call: bool,
}

#[derive(Debug, Deserialize)]
pub struct LibraryType {
    pub name: String,
    pub expr: String,
}

#[derive(Debug, Deserialize)]
pub struct LibraryConfig {
    pub headers: Vec<String>,
    pub nullopt: CallableLibraryType,
    pub optional: LibraryType,
    pub someopt: CallableLibraryType,
    pub span: LibraryType,
    pub string_view: LibraryType,
    pub string_view16: LibraryType,
    pub unique_ptr: LibraryType,
}

impl LibraryConfig {
    pub fn default() -> LibraryConfig {
        LibraryConfig {
            headers: vec!["#include <optional>".into()],
            nullopt: CallableLibraryType {
                name: "std::nullopt".into(),
                expr: "std::nullopt".into(),
                is_call: false,
            },
            optional: LibraryType {
                name: "std::optional".into(),
                expr: "std::optional".into(),
            },
            someopt: CallableLibraryType {
                name: "".into(),
                expr: "".into(),
                is_call: false,
            },
            span: LibraryType {
                // This is std::span compatible, and can be replaced with std::span
                // if targeting C++20. Internally the header will `using span = std::span`
                // anyway
                name: "diplomat::span".into(),
                expr: "diplomat::span".into(),
            },
            string_view: LibraryType {
                name: "std::string_view".into(),
                expr: "std::string_view".into(),
            },
            string_view16: LibraryType {
                name: "std::u16string_view".into(),
                expr: "std::u16string_view".into(),
            },
            unique_ptr: LibraryType {
                name: "std::unique_ptr".into(),
                expr: "std::unique_ptr".into(),
            },
        }
    }
}
