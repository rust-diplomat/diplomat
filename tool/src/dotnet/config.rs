use serde::Deserialize;

#[derive(Debug, Deserialize)]
pub struct LibraryConfig {
    pub namespace: String,
    pub native_lib: String,
    #[serde(default)]
    pub usings: Vec<String>,
    #[serde(default)]
    pub exceptions: ExceptionsConfig,
    #[serde(default)]
    pub properties: PropertiesConfig,
}

impl LibraryConfig {
    pub fn default() -> LibraryConfig {
        LibraryConfig {
            namespace: "Interop".to_owned(),
            native_lib: "rust".to_owned(),
            usings: Vec::new(),
            exceptions: ExceptionsConfig::default(),
            properties: PropertiesConfig::default(),
        }
    }
}

#[derive(Debug, Deserialize, Default)]
pub struct PropertiesConfig {
    #[serde(default)]
    pub setters_prefix: Option<String>,
    #[serde(default)]
    pub getters_prefix: Option<String>,
}

#[derive(Debug, Deserialize, Default)]
pub struct ExceptionsConfig {
    #[serde(default)]
    pub trim_suffix: String,
    #[serde(default)]
    pub error_message_method: Option<String>,
}
