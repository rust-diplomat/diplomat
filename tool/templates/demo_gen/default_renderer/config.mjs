// Don't really love this, but it works for now.
import cfg from "../diplomat.config.mjs"
export default cfg;