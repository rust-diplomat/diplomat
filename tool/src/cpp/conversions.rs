use std::fmt::Write;

use diplomat_core::ast;
use diplomat_core::Env;

use crate::cpp::config::LibraryConfig;

pub fn gen_rust_to_cpp<W: Write>(
    cpp: &str,
    path: &str,
    typ: &ast::TypeName,
    in_path: &ast::Path,
    env: &Env,
    library_config: &LibraryConfig,
    out: &mut W,
) -> String {
    match typ {
        ast::TypeName::Box(underlying) => match underlying.as_ref() {
            ast::TypeName::Named(path_type) | ast::TypeName::SelfType(path_type) => {
                match path_type.resolve(in_path, env) {
                    ast::CustomType::Opaque(opaque) => {
                        format!("{}({})", opaque.name, cpp)
                    }

                    ast::CustomType::Struct(_strct) => {
                        // TODO(#59): should emit a unique_ptr
                        todo!("Receiving boxes of structs is not yet supported")
                    }

                    ast::CustomType::Enum(_) => {
                        // TODO(#59): should emit a unique_ptr
                        todo!("Receiving boxes of enums is not yet supported")
                    }
                    &_ => unreachable!("unknown AST/HIR variant"),
                }
            }
            _o => todo!(),
        },
        ast::TypeName::Named(path_type) | ast::TypeName::SelfType(path_type) => {
            match path_type.resolve_with_path(in_path, env) {
                (_, ast::CustomType::Opaque(_)) => {
                    panic!("Cannot handle opaque structs by value");
                }

                (in_path, ast::CustomType::Struct(strct)) => {
                    let raw_struct_id = format!("diplomat_raw_struct_{path}");
                    writeln!(out, "capi::{} {} = {};", strct.name, raw_struct_id, cpp).unwrap();
                    let mut all_fields_wrapped = vec![];
                    for (name, typ, _) in &strct.fields {
                        all_fields_wrapped.push(format!(
                            ".{} = std::move({})",
                            name,
                            gen_rust_to_cpp(
                                &format!("{raw_struct_id}.{name}"),
                                &format!("{path}_{name}"),
                                typ,
                                &in_path,
                                env,
                                library_config,
                                out
                            )
                        ));
                    }

                    format!("{}{{ {} }}", strct.name, all_fields_wrapped.join(", "))
                }

                (_, ast::CustomType::Enum(enm)) => {
                    format!("static_cast<{}>({})", enm.name, cpp)
                }
                (_, &_) => unreachable!("unknown AST/HIR variant"),
            }
        }

        ast::TypeName::Option(underlying) => match underlying.as_ref() {
            ast::TypeName::Box(_) => {
                let raw_value_id = format!("diplomat_optional_raw_{path}");
                writeln!(out, "auto {raw_value_id} = {cpp};").unwrap();

                let ty_name =
                    super::types::gen_type(typ, in_path, None, env, library_config, false).unwrap();

                let wrapped_value_id = format!("diplomat_optional_{path}");
                writeln!(out, "{ty_name} {wrapped_value_id};").unwrap();

                writeln!(out, "if ({raw_value_id} != nullptr) {{").unwrap();

                let some_expr = gen_rust_to_cpp(
                    &raw_value_id,
                    path,
                    underlying.as_ref(),
                    in_path,
                    env,
                    library_config,
                    out,
                );
                if library_config.someopt.is_call {
                    writeln!(
                        out,
                        "  {} = {}({});",
                        wrapped_value_id, library_config.someopt.expr, some_expr
                    )
                    .unwrap();
                } else {
                    writeln!(
                        out,
                        "  {} = {}{};",
                        wrapped_value_id, library_config.someopt.expr, some_expr
                    )
                    .unwrap();
                }

                writeln!(out, "}} else {{").unwrap();
                if library_config.nullopt.is_call {
                    writeln!(
                        out,
                        "  {} = {}();",
                        wrapped_value_id, library_config.nullopt.expr
                    )
                    .unwrap();
                } else {
                    writeln!(
                        out,
                        "  {} = {};",
                        wrapped_value_id, library_config.nullopt.expr
                    )
                    .unwrap();
                }
                writeln!(out, "}}").unwrap();

                wrapped_value_id
            }

            _ => todo!(),
        },

        ast::TypeName::Result(ok, err, _) => {
            let raw_value_id = format!("diplomat_result_raw_{path}");
            writeln!(out, "auto {raw_value_id} = {cpp};").unwrap();
            let wrapped_value_id = format!("diplomat_result_{path}");
            let result_ty =
                super::types::gen_type(typ, in_path, None, env, library_config, false).unwrap();
            writeln!(out, "{result_ty} {wrapped_value_id};").unwrap();

            writeln!(out, "if ({raw_value_id}.is_ok) {{").unwrap();
            if !ok.is_zst() {
                let ok_expr = gen_rust_to_cpp(
                    &format!("{raw_value_id}.ok"),
                    path,
                    ok,
                    in_path,
                    env,
                    library_config,
                    out,
                );
                let ok_type =
                    super::types::gen_type(ok, in_path, None, env, library_config, false).unwrap();
                writeln!(
                    out,
                    "  {wrapped_value_id} = diplomat::Ok<{ok_type}>({ok_expr});"
                )
                .unwrap();
            } else {
                writeln!(
                    out,
                    "  {wrapped_value_id} = diplomat::Ok(std::monostate());"
                )
                .unwrap();
            };
            writeln!(out, "}} else {{").unwrap();

            if !err.is_zst() {
                let err_expr = gen_rust_to_cpp(
                    &format!("{raw_value_id}.err"),
                    path,
                    err,
                    in_path,
                    env,
                    library_config,
                    out,
                );
                let err_type =
                    super::types::gen_type(err, in_path, None, env, library_config, false).unwrap();
                writeln!(
                    out,
                    "  {wrapped_value_id} = diplomat::Err<{err_type}>({err_expr});"
                )
                .unwrap();
            } else {
                writeln!(
                    out,
                    "  {wrapped_value_id} = diplomat::Err(std::monostate());"
                )
                .unwrap();
            };
            writeln!(out, "}}").unwrap();

            wrapped_value_id
        }

        ast::TypeName::Primitive(_) => cpp.to_string(),
        ast::TypeName::Reference(_, _, _) => {
            todo!("Returning references from Rust to C++ is not currently supported")
        }
        ast::TypeName::Writeable => panic!("Returning writeables is not supported"),
        ast::TypeName::StrReference(
            _,
            ast::StringEncoding::UnvalidatedUtf8 | ast::StringEncoding::Utf8,
        ) => {
            let raw_value_id = format!("diplomat_str_raw_{path}");
            writeln!(out, "capi::DiplomatStringView {raw_value_id} = {cpp};").unwrap();

            writeln!(
                out,
                "std::string_view str({raw_value_id}.data, {raw_value_id}.len);"
            )
            .unwrap();
            "str".into()
        }
        ast::TypeName::StrReference(_, ast::StringEncoding::UnvalidatedUtf16) => {
            let raw_value_id = format!("diplomat_slice_raw_{path}");
            writeln!(out, "capi::DiplomatU16StringView {raw_value_id} = {cpp};").unwrap();

            let span = &library_config.span.expr;
            writeln!(
                out,
                "{span}<const char16_t> slice({raw_value_id}.data, {raw_value_id}.len);"
            )
            .unwrap();
            "slice".into()
        }
        ast::TypeName::PrimitiveSlice(_lt, mutability, prim) => {
            assert!(mutability.is_immutable());
            let raw_value_id = format!("diplomat_slice_raw_{path}");
            let mut prim_caps = prim.to_string();
            prim_caps.get_mut(0..1).unwrap().make_ascii_uppercase();
            let span = &library_config.span.expr;
            let prim = crate::c::types::c_type_for_prim(prim);
            writeln!(out, "capi::Diplomat{prim_caps}View {raw_value_id} = {cpp};").unwrap();

            writeln!(
                out,
                "{span}<const {prim}> slice({raw_value_id}.data, {raw_value_id}.len);"
            )
            .unwrap();
            "slice".into()
        }
        ast::TypeName::Unit => cpp.to_string(),
        &_ => unreachable!("unknown AST/HIR variant"),
    }
}

/// Meta information about a [ast::TypeName::Reference].
#[derive(Eq, PartialEq)]
pub struct ReferenceMeta {
    /// Whether or not the reference is owned.
    owned: bool,
    /// Whether or not the reference is mutable.
    mutable: bool,
    /// Whether the reference is nullable.
    is_nullable: bool,
}

#[allow(clippy::too_many_arguments)]
pub fn gen_cpp_to_rust<W: Write>(
    cpp: &str,
    path: &str,
    behind_ref: Option<ReferenceMeta>,
    typ: &ast::TypeName,
    in_path: &ast::Path,
    env: &Env,
    is_self: bool,
    out: &mut W,
) -> String {
    match typ {
        ast::TypeName::Reference(_, mutability, underlying) => {
            let behind_ref = match behind_ref {
                Some(mut br) => {
                    br.mutable = mutability.is_mutable();
                    br
                }
                None => ReferenceMeta {
                    owned: false,
                    mutable: mutability.is_mutable(),
                    is_nullable: false,
                },
            };
            gen_cpp_to_rust(
                cpp,
                path,
                Some(behind_ref),
                underlying.as_ref(),
                in_path,
                env,
                is_self,
                out,
            )
        }
        ast::TypeName::Named(path_type) | ast::TypeName::SelfType(path_type) => {
            match path_type.resolve(in_path, env) {
                ast::CustomType::Opaque(_opaque) => {
                    if let Some(reference) = behind_ref {
                        if is_self {
                            format!("{cpp}->inner.get()")
                        } else if reference.is_nullable {
                            format!("({cpp}) ? {cpp}->AsFFI() : nullptr")
                        } else if reference.mutable {
                            format!("{cpp}.AsFFIMut()")
                        } else {
                            format!("{cpp}.AsFFI()")
                        }
                    } else {
                        panic!("Cannot handle opaque types by value");
                    }
                }

                ast::CustomType::Struct(strct) => {
                    if let Some(reference) = behind_ref {
                        if reference.owned {
                            format!("(capi::{}*) {}", strct.name, cpp)
                        } else {
                            format!("(capi::{}*) &{}", strct.name, cpp)
                        }
                    } else {
                        let wrapped_struct_id = format!("diplomat_wrapped_struct_{path}");
                        writeln!(out, "{} {} = {};", strct.name, wrapped_struct_id, cpp).unwrap();
                        let mut all_fields_wrapped = vec![];
                        for (name, typ, _) in &strct.fields {
                            all_fields_wrapped.push(format!(
                                ".{} = {}",
                                name,
                                gen_cpp_to_rust(
                                    &format!("{wrapped_struct_id}.{name}"),
                                    &format!("{path}_{name}"),
                                    None,
                                    typ,
                                    in_path,
                                    env,
                                    false,
                                    out
                                )
                            ));
                        }

                        format!(
                            "capi::{}{{ {} }}",
                            strct.name,
                            all_fields_wrapped.join(", ")
                        )
                    }
                }

                ast::CustomType::Enum(enm) => format!("static_cast<capi::{}>({})", enm.name, cpp),
                &_ => unreachable!("unknown AST/HIR variant"),
            }
        }
        ast::TypeName::Writeable => {
            if behind_ref
                == Some(ReferenceMeta {
                    owned: false,
                    mutable: true,
                    is_nullable: false,
                })
            {
                writeln!(out, "capi::DiplomatWriteable {cpp}_writer = diplomat::WriteableTrait<W>::Construct({cpp});").unwrap();
                format!("&{cpp}_writer")
            } else {
                panic!("Cannot send Writeable to Rust as a value");
            }
        }
        ast::TypeName::Primitive(_) => cpp.to_string(),
        ast::TypeName::StrReference(..) => {
            format!("{{ {cpp}.data(), {cpp}.size() }}")
        }
        ast::TypeName::PrimitiveSlice(..) => {
            format!("{{ {cpp}.data(), {cpp}.size() }}")
        }
        ast::TypeName::Option(boxed) => match &**boxed {
            ast::TypeName::Reference(_, mutability, _) => {
                let behind_ref = match behind_ref {
                    Some(mut br) => {
                        br.mutable = mutability.is_mutable();
                        br.is_nullable = true;
                        br
                    }
                    None => ReferenceMeta {
                        owned: false,
                        mutable: mutability.is_mutable(),
                        is_nullable: true,
                    },
                };
                gen_cpp_to_rust(
                    cpp,
                    path,
                    Some(behind_ref),
                    boxed,
                    in_path,
                    env,
                    is_self,
                    out,
                )
            }
            o => todo!("{:?}", o),
        },
        o => todo!("{:?}", o),
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_enum_conversion() {
        test_file! {
            #[diplomat::bridge]
            mod ffi {
                enum MyEnum {
                    A, B, C
                }
                struct MyStruct {
                    a: u8,
                    b: MyEnum,
                }

                #[diplomat::opaque]
                struct Foo(Box<u8>);

                impl Foo {
                    pub fn get_struct(&self) -> MyStruct {
                        MyStruct { a: 1, b: MyEnum::A }
                    }
                }
            }
        }
    }

    #[test]
    fn test_option_conversion() {
        test_file! {
            #[diplomat::bridge]
            mod ffi {
                #[diplomat::opaque]
                struct MyStruct {
                    a: u8,
                }

                impl MyStruct {
                    pub fn create(&self) -> Option<Box<MyStruct>> {
                        unimplemented!();
                    }
                }
            }
        }
    }

    #[test]
    fn test_option_conversion_using_library_config() {
        test_file_using_library_config! {
            #[diplomat::bridge]
            mod ffi {
                #[diplomat::opaque]
                struct MyStruct {
                    a: u8,
                }

                impl MyStruct {
                    pub fn create(&self) -> Option<Box<MyStruct>> {
                        unimplemented!();
                    }
                }
            }
        }
    }
}
