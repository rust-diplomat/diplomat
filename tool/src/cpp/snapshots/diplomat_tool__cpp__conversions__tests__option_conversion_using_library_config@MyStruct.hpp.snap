---
source: tool/src/cpp/conversions.rs
expression: out_texts.get(out).unwrap()

---
#ifndef MyStruct_HPP
#define MyStruct_HPP
#include <stdint.h>
#include <stddef.h>
#include <stdbool.h>
#include <algorithm>
#include <memory>
#include <variant>

#include "mozilla/Maybe.h"
#include "mozilla/UniquePtr.h"
#include "mozilla/Span.h"
#include "diplomat_runtime.hpp"

namespace capi {
#include "MyStruct.h"
}

class MyStruct;

/**
 * A destruction policy for using MyStruct with mozilla::UniquePtr.
 */
struct MyStructDeleter {
  void operator()(capi::MyStruct* l) const noexcept {
    capi::MyStruct_destroy(l);
  }
};
class MyStruct {
 public:
  mozilla::Maybe<MyStruct> create() const;
  inline const capi::MyStruct* AsFFI() const { return this->inner.get(); }
  inline capi::MyStruct* AsFFIMut() { return this->inner.get(); }
  inline MyStruct(capi::MyStruct* i) : inner(i) {}
  MyStruct() = default;
  MyStruct(MyStruct&&) noexcept = default;
  MyStruct& operator=(MyStruct&& other) noexcept = default;
 private:
  mozilla::UniquePtr<capi::MyStruct, MyStructDeleter> inner;
};


inline mozilla::Maybe<MyStruct> MyStruct::create() const {
  auto diplomat_optional_raw_out_value = capi::MyStruct_create(this->inner.get());
  mozilla::Maybe<MyStruct> diplomat_optional_out_value;
  if (diplomat_optional_raw_out_value != nullptr) {
    diplomat_optional_out_value = mozilla::Some(MyStruct(diplomat_optional_raw_out_value));
  } else {
    diplomat_optional_out_value = mozilla::Nothing();
  }
  return diplomat_optional_out_value;
}
#endif

